lightning/kv/session.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kv
+
+import (
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	kvec "github.com/pingcap/tidb/util/kvencoder"
+)
+
+// transaction is a trimmed down Transaction type which only supports adding a
+// new KV pair.
+type transaction struct {
+	kv.Transaction
+	kvPairs []kvec.KvPair
+}
+
+// Set implements the kv.Transaction interface
+func (t *transaction) Set(k kv.Key, v []byte) error {
+	t.kvPairs = append(t.kvPairs, kvec.KvPair{
+		Key: k.Clone(),
+		Val: append([]byte{}, v...),
+	})
+	return nil
+}
+
+// SetOption implements the kv.Transaction interface
+func (t *transaction) SetOption(opt kv.Option, val interface{}) {}
+
+// DelOption implements the kv.Transaction interface
+func (t *transaction) DelOption(kv.Option) {}
+
+// transaction is a trimmed down Transaction type which only supports adding a
+// new KV pair.
+type session struct {
+	sessionctx.Context
+	txn  transaction
+	vars *variable.SessionVars
+}
+
+func newSession(sqlMode mysql.SQLMode) *session {
+	vars := variable.NewSessionVars()
+	vars.LightningMode = true
+	vars.SkipUTF8Check = true
+	vars.StmtCtx.InInsertStmt = true
+	vars.StmtCtx.BadNullAsWarning = !sqlMode.HasStrictMode()
+	vars.StmtCtx.TruncateAsWarning = !sqlMode.HasStrictMode()
+	vars.StmtCtx.OverflowAsWarning = !sqlMode.HasStrictMode()
+	vars.StmtCtx.AllowInvalidDate = sqlMode.HasAllowInvalidDatesMode()
+	vars.StmtCtx.IgnoreZeroInDate = !sqlMode.HasStrictMode() || sqlMode.HasAllowInvalidDatesMode()
+	vars.StmtCtx.TimeZone = vars.Location()
+	return &session{
+		txn:  transaction{},
+		vars: vars,
+	}
+}
+
+func (se *session) takeKvPairs() []kvec.KvPair {
+	pairs := se.txn.kvPairs
+	se.txn.kvPairs = make([]kvec.KvPair, 0, len(pairs))
```

Comment:
Maybe we could use a memory pool to reduce memory allocation.

Reply:
one question: if we want to reduce memory allocation, is it the best way to put `kvPair` into memory pool?

Reply:
This is a much bigger change than that record cache, and since the KV pairs are shared across threads it may be harder to ensure thread safety. I'd like to hold this off unless the memory profile shows otherwise.

Reply:
Actually, I think it is better to maintain a free list in session struct.

Reply:
@kennytm Yes, it's OK for lightning.